[BUG] Failed to get usage limit info(v2)

Docker image pulling time:
`2025-09-19 20:00 UTC`
Source of Docker image:
`ghcr.io/simstudioai/simstudio:latest`

After local deployment via Docker, the Block cannot be used. Clicking on an item in the Blocks area with the mouse or dragging the target item to the editing area has no response. The following is the simstudio log when the problem occurs:
```log
   ▲ Next.js 15.4.1

   - Local:        http://localhost:3000

   - Network:      http://0.0.0.0:3000



 ✓ Starting...

[Main Instrumentation] register() called, environment: {

  NEXT_RUNTIME: "nodejs",

  NODE_ENV: "production",

}

[Main Instrumentation] Loading Node.js instrumentation...

 ✓ Ready in 250ms

[Main Instrumentation] Calling Node.js register()...

2025-09-20T14:15:57.921Z WARN [Better Auth]: Social provider github is missing clientId or clientSecret

2025-09-20T14:15:57.921Z WARN [Better Auth]: Social provider google is missing clientId or clientSecret

2025-09-20T14:16:02.178Z WARN [Better Auth]: Social provider github is missing clientId or clientSecret

2025-09-20T14:16:02.178Z WARN [Better Auth]: Social provider google is missing clientId or clientSecret

[Main Instrumentation] register() called, environment: { NEXT_RUNTIME: 'edge', NODE_ENV: 'production' }

[Main Instrumentation] Loading Edge Runtime instrumentation...

[Main Instrumentation] Calling Edge Runtime register()...

[2025-09-20T14:16:17.367Z] [ERROR] [UsageManagement] Failed to get usage limit info {"userId":"11223344556677889900","error":{}}

[2025-09-20T14:16:17.367Z] [ERROR] [UnifiedUsageAPI] Failed to get usage limit info {"userId":"11223344556677889900","error":{}}

[2025-09-20T14:16:17.828Z] [ERROR] [OllamaModelsAPI] Failed to fetch Ollama models {"error":"Unable to connect. Is the computer able to access the url?","host":"http://localhost:11434"}
```

The following is my Docker Compose configuration:
```yaml
networks:
  default:
    external: true
    name: local

services:
  simstudio:
    image: ghcr.io/simstudioai/simstudio:latest
    restart: unless-stopped
    container_name: simstudio
    ports:
      - '3000:3000'
    environment:
      - DATABASE_URL=postgresql://${POSTGRES_USER:-postgres}:${POSTGRES_PASSWORD:-postgres}@postgresql:5432/${POSTGRES_DB:-simstudio}
      - BETTER_AUTH_URL=${NEXT_PUBLIC_APP_URL:-https://myDomain:3000}
      - NEXT_PUBLIC_APP_URL=${NEXT_PUBLIC_APP_URL:-https://myDomain:3000}
      - BETTER_AUTH_SECRET=${BETTER_AUTH_SECRET:-11223344556677889900}
      - ENCRYPTION_KEY=${ENCRYPTION_KEY:-11223344556677889900}
      # - OLLAMA_URL=${OLLAMA_URL:-http://localhost:11434}
      - SOCKET_SERVER_URL=${SOCKET_SERVER_URL:-http://realtime:3002}
      - NEXT_PUBLIC_SOCKET_URL=${NEXT_PUBLIC_SOCKET_URL:-http://realtime:3002}
      - RESEND_API_KEY=re_11223344556677889900
      - EMAIL_DOMAIN=notifications.myDomain
    depends_on:
      realtime:
        condition: service_healthy
    healthcheck:
      test: ['CMD', 'wget', '--spider', '--quiet', 'http://127.0.0.1:3000']
      interval: 90s
      timeout: 5s
      retries: 3
      start_period: 10s
    deploy:
      resources:
        limits:
          memory: 2G
        reservations:
          memory: 512MB

  realtime:
    image: ghcr.io/simstudioai/realtime:latest
    restart: unless-stopped
    container_name: realtime
    ports:
      - '3002:3002'
    environment:
      - DATABASE_URL=postgresql://${POSTGRES_USER:-postgres}:${POSTGRES_PASSWORD:-postgres}@postgresql:5432/${POSTGRES_DB:-simstudio}
      - NEXT_PUBLIC_APP_URL=${NEXT_PUBLIC_APP_URL:-http://simstudio:3000}
      - BETTER_AUTH_URL=${BETTER_AUTH_URL:-http://simstudio:3000}
      - BETTER_AUTH_SECRET=${BETTER_AUTH_SECRET:-11223344556677889900}
    healthcheck:
      test: ['CMD', 'wget', '--spider', '--quiet', 'http://127.0.0.1:3002/health']
      interval: 90s
      timeout: 5s
      retries: 3
      start_period: 10s
    deploy:
      resources:
        limits:
          memory: 512G
        reservations:
          memory: 128MB
```

To prevent the warning "Some containers have not started" from appearing in my Docker graphical interface, I separated migrations from Compose:
```yaml
networks:
  default:
    external: true
    name: local

services:
  simstudio-migrations:
    image: ghcr.io/simstudioai/migrations:latest
    working_dir: /app/packages/db
    container_name: simstudio-migrations
    environment:
      - DATABASE_URL=postgresql://${POSTGRES_USER:-postgres}:${POSTGRES_PASSWORD:-postgres}@postgresql:5432/${POSTGRES_DB:-simstudio}
    command: ['bun', 'run', 'db:migrate']
    restart: 'no'
    deploy:
      resources:
        limits:
          memory: 1G
        reservations:
          memory: 256MB
```

The following is my PostgreSQL configuration:

```yaml
networks:
  default:
    external: true
    name: local

services:
  postgresql:
    image: pgvector/pgvector:pg17
    restart: always
    container_name: postgresql
    ports:
      - '${POSTGRES_PORT:-5432}:5432'
    environment:
      - POSTGRES_USER=${POSTGRES_USER:-postgres}
      - POSTGRES_PASSWORD=${POSTGRES_PASSWORD:-postgres}
      - POSTGRES_DB=${POSTGRES_DB:-simstudio}
    volumes:
      - ./data:/var/lib/postgresql/data
    healthcheck:
      test: ['CMD-SHELL', 'pg_isready -U postgres']
      interval: 5s
      timeout: 5s
      retries: 5
    deploy:
      resources:
        limits:
          memory: 512MB
        reservations:
          memory: 72MB
```

The following is the log of realtime:
```log
2025-09-20T14:15:51.955Z WARN [Better Auth]: Social provider github is missing clientId or clientSecret

2025-09-20T14:15:51.956Z WARN [Better Auth]: Social provider google is missing clientId or clientSecret
```

The following is the log of migrations:
```log
$ bunx drizzle-kit migrate --config=./drizzle.config.ts

Reading config file '/app/packages/db/drizzle.config.ts'

Using 'postgres' driver for database querying

[⣷] applying migrations...{

  severity_local: "NOTICE",

  severity: "NOTICE",

  code: "42P06",

  message: "schema \"drizzle\" already exists, skipping",

  file: "schemacmds.c",

  line: "132",

  routine: "CreateSchemaCommand",

}

{

  severity_local: "NOTICE",

  severity: "NOTICE",

  code: "42P07",

  message: "relation \"__drizzle_migrations\" already exists, skipping",

  file: "parse_utilcmd.c",

  line: "207",

  routine: "transformCreateStmt",

}

[✓] migrations applied successfully!
```

I replaced the sensitive information with meaningless strings.

My service is deployed in mainland China, and I suspect the Failed to fetch error might be related to network restrictions.
Hope this issue can be resolved.

--- a/apps/sim/lib/auth.ts
+++ b/apps/sim/lib/auth.ts
@@ -73,6 +73,24 @@ export const auth = betterAuth({
     freshAge: 60 * 60, // 1 hour (or set to 0 to disable completely)
   },
   databaseHooks: {
+    user: {
+      create: {
+        after: async (user) => {
+          logger.info('[databaseHooks.user.create.after] User created, initializing stats', {
+            userId: user.id,
+          })
+
+          try {
+            await handleNewUser(user.id)
+          } catch (error) {
+            logger.error('[databaseHooks.user.create.after] Failed to initialize user stats', {
+              userId: user.id,
+              error,
+            })
+          }
+        },
+      },
+    },
     session: {
       create: {
         before: async (session) => {
@@ -1152,15 +1170,6 @@ export const auth = betterAuth({
                 stripeCustomerId: stripeCustomer.id,
                 userId: user.id,
               })
-
-              try {
-                await handleNewUser(user.id)
-              } catch (error) {
-                logger.error('[onCustomerCreate] Failed to handle new user setup', {
-                  userId: user.id,
-                  error,
-                })
-              }
             },
             subscription: {
               enabled: true,

--- a/apps/sim/lib/billing/calculations/usage-monitor.ts
+++ b/apps/sim/lib/billing/calculations/usage-monitor.ts
@@ -8,7 +8,6 @@ import { createLogger } from '@/lib/logs/console/logger'
 
 const logger = createLogger('UsageMonitor')
 
-// Percentage threshold for showing warning
 const WARNING_THRESHOLD = 80
 
 interface UsageData {
@@ -157,13 +156,18 @@ export async function checkUsageStatus(userId: string): Promise<UsageData> {
       userId,
     })
 
-    // Return default values in case of error
+    // Block execution if we can't determine usage status
+    logger.error('Cannot determine usage status - blocking execution', {
+      userId,
+      error: error instanceof Error ? error.message : String(error),
+    })
+
     return {
-      percentUsed: 0,
+      percentUsed: 100,
       isWarning: false,
-      isExceeded: false,
+      isExceeded: true, // Block execution when we can't determine status
       currentUsage: 0,
-      limit: 0,
+      limit: 0, // Zero limit forces blocking
     }
   }
 }
@@ -241,7 +245,6 @@ export async function checkServerSideUsageLimits(userId: string): Promise<{
   message?: string
 }> {
   try {
-    // If billing is disabled, always allow execution
     if (!isBillingEnabled) {
       return {
         isExceeded: false,
@@ -252,7 +255,6 @@ export async function checkServerSideUsageLimits(userId: string): Promise<{
 
     logger.info('Server-side checking usage limits for user', { userId })
 
-    // Hard block if billing is flagged as blocked
     const stats = await db
       .select({
         blocked: userStats.billingBlocked,
@@ -274,7 +276,6 @@ export async function checkServerSideUsageLimits(userId: string): Promise<{
       }
     }
 
-    // Get usage data using the same function we use for client-side
     const usageData = await checkUsageStatus(userId)
 
     return {
@@ -291,12 +292,19 @@ export async function checkServerSideUsageLimits(userId: string): Promise<{
       userId,
     })
 
-    // Be conservative in case of error - allow execution but log the issue
+    logger.error('Cannot determine usage limits - blocking execution', {
+      userId,
+      error: error instanceof Error ? error.message : String(error),
+    })
+
     return {
-      isExceeded: false,
+      isExceeded: true, // Block execution when we can't determine limits
       currentUsage: 0,
-      limit: 0,
-      message: `Error checking usage limits: ${error instanceof Error ? error.message : String(error)}`,
+      limit: 0, // Zero limit forces blocking
+      message:
+        error instanceof Error && error.message.includes('No user stats record found')
+          ? 'User account not properly initialized. Please contact support.'
+          : 'Unable to determine usage limits. Execution blocked for security. Please contact support.',
     }
   }
 }

--- a/apps/sim/lib/billing/core/usage.ts
+++ b/apps/sim/lib/billing/core/usage.ts
@@ -22,18 +22,6 @@ const logger = createLogger('UsageManagement')
  */
 export async function handleNewUser(userId: string): Promise<void> {
   try {
-    // Check if user stats already exists to make this function idempotent
-    const existingStats = await db
-      .select({ id: userStats.id })
-      .from(userStats)
-      .where(eq(userStats.userId, userId))
-      .limit(1)
-
-    if (existingStats.length > 0) {
-      logger.info('User stats record already exists for user', { userId })
-      return
-    }
-
     await db.insert(userStats).values({
       id: crypto.randomUUID(),
       userId: userId,

--- a/apps/sim/lib/billing/core/usage.ts
+++ b/apps/sim/lib/billing/core/usage.ts
@@ -22,6 +22,18 @@ const logger = createLogger('UsageManagement')
  */
 export async function handleNewUser(userId: string): Promise<void> {
   try {
+    // Check if user stats already exists to make this function idempotent
+    const existingStats = await db
+      .select({ id: userStats.id })
+      .from(userStats)
+      .where(eq(userStats.userId, userId))
+      .limit(1)
+
+    if (existingStats.length > 0) {
+      logger.info('User stats record already exists for user', { userId })
+      return
+    }
+
     await db.insert(userStats).values({
       id: crypto.randomUUID(),
       userId: userId,
@@ -312,13 +324,15 @@ export async function getUserUsageLimit(userId: string): Promise<number> {
       .limit(1)
 
     if (userStatsQuery.length === 0) {
-      throw new Error(`User stats not found for userId: ${userId}`)
+      throw new Error(
+        `No user stats record found for userId: ${userId}. User must be properly initialized before execution.`
+      )
     }
 
     // Individual limits should never be null for free/pro users
     if (!userStatsQuery[0].currentUsageLimit) {
       throw new Error(
-        `Invalid null usage limit for ${subscription?.plan || 'free'} user: ${userId}`
+        `Invalid null usage limit for ${subscription?.plan || 'free'} user: ${userId}. User stats must be properly initialized.`
       )
     }
 
@@ -332,7 +346,7 @@ export async function getUserUsageLimit(userId: string): Promise<number> {
     .limit(1)
 
   if (orgData.length === 0) {
-    throw new Error(`Organization not found: ${subscription.referenceId}`)
+    throw new Error(`Organization not found: ${subscription.referenceId} for user: ${userId}`)
   }
 
   if (orgData[0].orgUsageLimit) {

--- a/apps/sim/lib/logs/execution/logger.ts
+++ b/apps/sim/lib/logs/execution/logger.ts
@@ -403,52 +403,59 @@ export class ExecutionLogger implements IExecutionLoggerService {
       // Apply cost multiplier only to model costs, not base execution charge
       const costToStore = costSummary.baseExecutionCharge + costSummary.modelCost * costMultiplier
 
-      // Check if user stats record exists
-      const userStatsRecords = await db.select().from(userStats).where(eq(userStats.userId, userId))
+      // Upsert user stats record - insert if doesn't exist, update if it does
+      const { getFreeTierLimit } = await import('@/lib/billing/subscriptions/utils')
+      const defaultLimit = getFreeTierLimit()
 
-      if (userStatsRecords.length > 0) {
-        // Update user stats record with trigger-specific increments
-        const updateFields: any = {
-          totalTokensUsed: sql`total_tokens_used + ${costSummary.totalTokens}`,
-          totalCost: sql`total_cost + ${costToStore}`,
-          currentPeriodCost: sql`current_period_cost + ${costToStore}`, // Track current billing period usage
-          lastActive: new Date(),
-        }
-
-        // Add trigger-specific increment
-        switch (trigger) {
-          case 'manual':
-            updateFields.totalManualExecutions = sql`total_manual_executions + 1`
-            break
-          case 'api':
-            updateFields.totalApiCalls = sql`total_api_calls + 1`
-            break
-          case 'webhook':
-            updateFields.totalWebhookTriggers = sql`total_webhook_triggers + 1`
-            break
-          case 'schedule':
-            updateFields.totalScheduledExecutions = sql`total_scheduled_executions + 1`
-            break
-          case 'chat':
-            updateFields.totalChatExecutions = sql`total_chat_executions + 1`
-            break
-        }
-
-        await db.update(userStats).set(updateFields).where(eq(userStats.userId, userId))
-
-        logger.debug('Updated user stats record with cost data', {
-          userId,
-          trigger,
-          addedCost: costToStore,
-          addedTokens: costSummary.totalTokens,
-        })
-      } else {
-        logger.error('User stats record not found - should be created during onboarding', {
-          userId,
-          trigger,
-        })
-        return // Skip cost tracking if user stats doesn't exist
+      const triggerIncrements: any = {}
+      switch (trigger) {
+        case 'manual':
+          triggerIncrements.totalManualExecutions = sql`total_manual_executions + 1`
+          break
+        case 'api':
+          triggerIncrements.totalApiCalls = sql`total_api_calls + 1`
+          break
+        case 'webhook':
+          triggerIncrements.totalWebhookTriggers = sql`total_webhook_triggers + 1`
+          break
+        case 'schedule':
+          triggerIncrements.totalScheduledExecutions = sql`total_scheduled_executions + 1`
+          break
+        case 'chat':
+          triggerIncrements.totalChatExecutions = sql`total_chat_executions + 1`
+          break
       }
+
+      await db
+        .insert(userStats)
+        .values({
+          id: uuidv4(),
+          userId: userId,
+          currentUsageLimit: defaultLimit.toString(),
+          usageLimitUpdatedAt: new Date(),
+          totalTokensUsed: costSummary.totalTokens,
+          totalCost: costToStore,
+          currentPeriodCost: costToStore,
+          lastActive: new Date(),
+          ...triggerIncrements,
+        })
+        .onConflictDoUpdate({
+          target: userStats.userId,
+          set: {
+            totalTokensUsed: sql`total_tokens_used + ${costSummary.totalTokens}`,
+            totalCost: sql`total_cost + ${costToStore}`,
+            currentPeriodCost: sql`current_period_cost + ${costToStore}`,
+            lastActive: new Date(),
+            ...triggerIncrements,
+          },
+        })
+
+      logger.debug('Upserted user stats record with cost data', {
+        userId,
+        trigger,
+        addedCost: costToStore,
+        addedTokens: costSummary.totalTokens,
+      })
     } catch (error) {
       logger.error('Error updating user stats with cost information', {
         workflowId,

